File: src/test/java/org/micoli/php/AttributeNavigationTest.java
package org.micoli.php;

import com.intellij.codeInsight.daemon.GutterMark;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import java.util.List;
import java.util.Objects;
import org.micoli.php.attributeNavigation.configuration.AttributeNavigationConfiguration;
import org.micoli.php.attributeNavigation.configuration.NavigationByAttributeRule;
import org.micoli.php.attributeNavigation.service.AttributeNavigationService;
import org.micoli.php.configuration.ConfigurationException;
import org.micoli.php.configuration.ConfigurationFactory;
import org.micoli.php.configuration.NoConfigurationFileException;

public class AttributeNavigationTest extends BasePlatformTestCase {

    @Override
    protected String getTestDataPath() {
        return "src/test/resources/testData";
    }

    public void testItFormatValueUsingInlineFormatter() {
        String formattedValue = AttributeNavigationService.getInstance(getProject())
                .getFormattedValue("cde", "return ('ab-'+value+'-fg').toLowerCase()");
        assertEquals("ab-cde-fg", formattedValue);
    }

    public void testItCanFindLineMarkersForAttributes() {
        myFixture.configureByFiles("src/UserInterface/Web/Api/Article/Get/Controller.php");
        loadPluginConfiguration(getTestDataPath());
        List<GutterMark> lineMarkers = myFixture.findAllGutters();
        assertNotEmpty(lineMarkers);

        List<GutterMark> specificMarkers = lineMarkers.stream()
                .filter(it -> {
                    String tooltipText = it.getTooltipText();
                    if (tooltipText == null) {
                        return false;
                    }
                    return tooltipText.contains("Search for [");
                })
                .toList();

        assertEquals(1, specificMarkers.size());
    }

    public void testItFormatValueUsingScriptInConfiguration() {
        AttributeNavigationService instance = loadPluginConfiguration(getTestDataPath());
        NavigationByAttributeRule rule = instance.getRules().getFirst();
        String formattedValue =
                instance.getFormattedValue("/templates/{templateId}/documents/{documentId}", rule.formatterScript);
        assertEquals("/templates/[^/]*/documents/[^/]*:", formattedValue);
    }

    private AttributeNavigationService loadPluginConfiguration(String path) {
        AttributeNavigationConfiguration attributeNavigationConfiguration = null;
        try {
            attributeNavigationConfiguration = Objects.requireNonNull(ConfigurationFactory.loadConfiguration(path, 0L))
                    .configuration
                    .attributeNavigation;
        } catch (ConfigurationException | NoConfigurationFileException e) {
            throw new RuntimeException(e);
        }
        AttributeNavigationService instance = AttributeNavigationService.getInstance(getProject());
        instance.loadConfiguration(myFixture.getProject(), attributeNavigationConfiguration);

        return instance;
    }
}

File: src/test/kotlin/org/micoli/php/AttributeNavigationTest.kt
package org.micoli.php

import com.intellij.testFramework.fixtures.BasePlatformTestCase
import org.assertj.core.api.Assertions.*
import org.micoli.php.attributeNavigation.configuration.AttributeNavigationConfiguration
import org.micoli.php.attributeNavigation.configuration.NavigationByAttributeRule
import org.micoli.php.attributeNavigation.service.AttributeNavigationService
import org.micoli.php.configuration.ConfigurationException
import org.micoli.php.configuration.ConfigurationFactory
import org.micoli.php.configuration.exceptions.NoConfigurationFileException

class AttributeNavigationTest : BasePlatformTestCase() {
    override fun getTestDataPath(): String = "src/test/resources/symfony-demo"

    fun testItFormatValueUsingInlineFormatter() {
        val formattedValue =
            AttributeNavigationService.getInstance(project)
                .getFormattedValue("cde", "return ('ab-'+value+'-fg').toLowerCase()")
        assertThat(formattedValue).isEqualTo("ab-cde-fg")
    }

    fun testItCanFindLineMarkersForAttributes() {
        myFixture.configureByFiles("src/Controller/BlogController.php")
        loadPluginConfiguration(testDataPath)
        val lineMarkers = myFixture.findAllGutters()
        assertThat(lineMarkers).isNotEmpty

        val specificMarkers =
            lineMarkers
                .stream()
                .map { it?.tooltipText?.replace("Search for ", "") }
                .toList()
                .filterNotNull()
                .sorted()
                .joinToString(",")

        assertThat(specificMarkers)
            .isEqualTo(
                "[/],[/blog],[/comment/{postSlug}/new],[/page/{page}],[/posts/{slug:post}],[/rss.xml],[/search]")
    }

    fun testItFormatValueUsingScriptInConfiguration() {
        val instance = loadPluginConfiguration(testDataPath)
        val rule: NavigationByAttributeRule = instance.rules.first()
        val formattedValue =
            instance.getFormattedValue(
                "/templates/{templateId}/documents/{documentId}", rule.formatterScript)
        assertEquals("/templates/[^/]*/documents/[^/]*:", formattedValue)
    }

    private fun loadPluginConfiguration(path: String): AttributeNavigationService {
        var attributeNavigationConfiguration: AttributeNavigationConfiguration?
        try {
            attributeNavigationConfiguration =
                ConfigurationFactory()
                    .loadConfiguration(path, 0L, true)
                    ?.configuration
                    ?.attributeNavigation
        } catch (e: ConfigurationException) {
            throw RuntimeException(e)
        } catch (e: NoConfigurationFileException) {
            throw RuntimeException(e)
        }
        val instance = AttributeNavigationService.getInstance(project)
        instance.loadConfiguration(attributeNavigationConfiguration)

        return instance
    }
}
